### DISCUSSION TITLE
Consider making 'sniffio' dependency optional

### DISCUSSION THREAD

**Discussion by rc-mattschwager** [encode/httpx#2260] [General]
Hi there!

Thank you for the excellent `httpx` library. I use it everyday, and it makes async programming in Python a joy.

In the interest of dependency minimization, I would like to purpose making the `sniffio` dependency/functionality optional. Correct me if I'm wrong here, but I suspect 99% of users are using the stdlib async implementation, so `sniffio` is unnecessary. Further, users leveraging something like `trio` or `curio` are likely power users who know what they're doing, and can jump through the additional hoop of optional dependencies.

`httpx` appears to have a history of dependency minimization (https://github.com/encode/httpx/pull/2165, https://github.com/encode/httpx/pull/2252), and making judicious use of optional dependencies (https://github.com/encode/httpx#dependencies). Further, `sniffio` usage is minimal, and thus shouldn't be too difficult to manage as an optional dependency:

- https://github.com/encode/httpx/blob/9baf3a6cd2/httpx/_transports/asgi.py#L18
- https://github.com/encode/httpx/blob/9baf3a6cd2/httpx/_utils.py#L387
- https://github.com/encode/httpx/blob/9baf3a6cd2/tests/concurrency.py#L12

Reviewing the original PR (https://github.com/encode/httpx/pull/585) it doesn't appear that there was any major driving factor in its inclusion, at least that I can see.

Let me know if I'm missing anything, and I'm happy to open the discussion on this subject.

**Comment by lovelydinosaur** [comment]
The `sniffio` package is a super tiny dependency, so the thing to do here would be to figure out how to replicate the auto-detection of "are we in an asyncio or a trio context", without using the package.

From a bit of digging I think we can do this ourselves, while still relying on public API...

```python
def determine_async_context():
    try:
        if asyncio.current_task() is not None:
            return 'asyncio'
    except RuntimeError:
        pass

    try:
        import trio
        if trio.lowlevel.current_task() is not None:
            return 'trio'
    except (RuntimeError, ImportError):
        pass

    raise RuntimeError("determine_async_context called, but no task is running")
```

I think that switching around to something like that could probably allow us to remove our `sniffio` dependancy.

As it happens, we've got a much more substantial async dependancy in the form of `anyio`, which `httpcore` uses for it's async implementation. However let's discuss that separately.

**Reply by rc-mattschwager** [reply]
Ahh, that's an interesting approach. I was envisioning something different, but that would work as well.

I was thinking something more along the lines of:

```python
try:
    import sniffio
except ImportError:
    sniffio = None

...

if sniffio is not None and sniffio.current_async_library() ...
    ...
```

Then there could be an optional dependency along the lines of `httpx[sniffio]`.

I think either approach would work well to achieve the spirit of the original request. What do you think of the tradeoffs between these two approaches?
